enhancements/microshift/microshift-router-configuration.md

```diff
@@ -0,0 +1,437 @@
+
+### Non-Goals
+* N/A
+
+## Proposal
+Each configuration option will get its own section with specifics and
+details.
+As a common baseline, a new top level section will be added to the
+configuration:
+```yaml
+apiServer:
+   ...
+# new from here.
+ingress:
+  status: <Enabled|Disabled> # Defaults to Enabled.
+  ports:
+    http: <int> # Defaults to 80.
+    https: <int> # Defaults to 443.
+  expose:
+    - <IP address/NIC> # Defaults to the NICs of the node
+```
+
+### Enable/disable the router
+The following configuration is proposed:
+```yaml
+ingress:
+  status: <Enabled|Disabled> # Defaults to Enabled.
+```
+
+With this option MicroShift will decide whether it should create the router
+upon starting. This includes not just the pod, but all the associated resources
+that come along with it: namespace, services, configmaps, etc.
+
+When setting the option to `Disabled`, the next MicroShift restart should delete
+all the router related resources, including APIs and controllers.
+
+Setting the option to `Enabled` will deploy the router, while `Disabled` will disable
+it. Default will be `Enabled`.
+
+### Listening ports
+The following configuration is proposed:
+```yaml
+ingress:
+  ports:
+    http: <int> # Defaults to 80.
+    https: <int> # Defaults to 443.
+```
+
+In order to allow this configuration and many other advantages, the
+router service shall be exposed using `LoadBalancer` type. See the
+Design Details section for more information.
+
+#### Firewalling ports
+Using `LoadBalancer` service type prevents the usage of firewalld to block
+access to ports. Since this is a feature that users may require for auditing,
+an alternative is provided.
+
+The `NetworkPolicy` resource fits the needs for blocking traffic to services
+and pods. Policies can also be configured to allow specific IP addresses and
+provide a high degree of customization for network traffic. They use the CNI
+instead of host level settings, providing greater flexibility and being self
+contained within MicroShift.
+
+Users in need of blocking traffic shall use the `NetworkPolicy` resource. A
+section in the documentation will be added to describe how to do this.
+
+### Configure IPs where the router listens
+The following configuration is proposed:
+```yaml
+ingress:
+  expose:
+    - <IP address/NIC> # Defaults to the NICs of the node
+```
+
+As we have seen in previous sections, the use of `LoadBalancer` makes ovnk
+configure iptables rules to expose the service outside of the cluster.
+Building on [LoadBalancer controller enhancement](loadbalancer-service-support.md),
+the service is exposed using the node IP and configured ports.
+
+All the IPs where the router is exposed need to be part of the `LoadBalancer`
+service status definition. In order to provide more flexibility, the
+configuration options are able to take IP addresses and/or NIC names in the
+system. Using NIC names abstracts away IP changes, in cases where this
+happens.
```

Comment:
Does having the router listen on multiple IPs have any effect on the certificates needed?

Reply:
The question about certificates is interesting in light of the `expose.hostnames` option that this EP adds.  With this new option, you can specify a host name that is entirely different from the cluster's base domain.  Is the ingress domain also configurable in MicroShift?  Is having all these different places you can configure a domain name or host name (or *must* configure them in order to keep the configuration consistent) going to be confusing or troublesome?